backend/app/core/config.py

```diff
@@ -13,10 +14,40 @@ class ServerSessionsSettings(BaseSettings):
     expires_in: int
 
 
+class MatrixBotSettings(BaseSettings):
+    homeserver: str
+    user_id: str
+    access_token: str
+
+    # This checks if the user_id and homeserver match the rules set by the matrix spec.
+    # This returns a dictionary instead of a MatrixBotSettings object
+    @root_validator(pre=True)
+    def validate(cls, values):
+        homeserver, user_id = values.get("homeserver"), values.get("user_id")
+        list_of_credentials = user_id.split(":")
+        if len(list_of_credentials) < 2:
+            raise ValueError("Invalid user_id")
+
+        username, user_homeserver = list_of_credentials
+        if re.search(r"@[\w\_\-]+", username) is None:
+            raise ValueError("Invalid user_id")
+
+        if re.sub(r"https?://", "", homeserver) != user_homeserver:
+            raise ValueError("Invalid homeserver or user_id")
```

Comment:
This is incorrect. There is no requirement that the domain part of the user id and the homeserver address be the same. See https://spec.matrix.org/v1.2/server-server-api/#resolving-server-names for all the *gory* details.